src/ripple/app/tx/impl/Transactor.cpp

```diff
@@ -688,10 +692,22 @@ Transactor::operator()()
 
     if (didApply)
     {
+        // Check invariants
+        // if `tecINVARIANT_FAILED` not returned, we can proceed to apply the tx
+        terResult = ctx_.checkInvariants(terResult);
+        if (terResult == tecINVARIANT_FAILED)
+        {
+            // if invariants failed, claim a fee still
+            claimFee(fee, terResult, {});
+            //Check invariants *again* to ensure the fee claiming doesn't
+            //violate invariants.
+            terResult = ctx_.checkInvariants(terResult);
```

Comment:
I had the mistaken impression that `apply()` would handle this - but I think you are correct about needing to guard against the tef - will fix.